File: apps/api/app/schemas/data_analytics_mlops.py
import uuid
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


# --- Dataset & Pipeline Schemas ---
class DatasetBase(BaseModel):
    dataset_name: str = Field(..., min_length=1, max_length=255)
    dataset_type: str = Field(default="Tabular", max_length=50)
    source: str = Field(..., min_length=1, max_length=255)
    schema_version: str = Field(default="v1.0", max_length=50)
    status: str = Field(default="Active", max_length=50)


class DatasetCreate(DatasetBase):
    organization_id: uuid.UUID


class DatasetResponse(DatasetBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PipelineJobBase(BaseModel):
    pipeline_name: str = Field(..., min_length=1, max_length=255)
    schedule_cron: str | None = Field(None, max_length=100)
    status: str = Field(default="Active", max_length=50)


class PipelineJobCreate(PipelineJobBase):
    organization_id: uuid.UUID


class PipelineJobResponse(PipelineJobBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ETLJobBase(BaseModel):
    job_name: str = Field(..., min_length=1, max_length=255)
    source_type: str = Field(..., min_length=1, max_length=100)
    target_type: str = Field(..., min_length=1, max_length=100)
    status: str = Field(default="Idle", max_length=50)


class ETLJobCreate(ETLJobBase):
    pass


class ETLJobResponse(ETLJobBase):
    id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Feature Store Schemas ---
class FeatureStoreBase(BaseModel):
    feature_name: str = Field(..., min_length=1, max_length=255)
    feature_group: str = Field(..., min_length=1, max_length=100)
    data_type: str = Field(default="FLOAT", max_length=50)
    description: str | None = Field(None, max_length=1000)
    version: str = Field(default="v1.0", max_length=50)
    status: str = Field(default="Active", max_length=50)


class FeatureStoreCreate(FeatureStoreBase):
    feature_group_id: uuid.UUID | None = None


class FeatureStoreResponse(FeatureStoreBase):
    id: uuid.UUID
    feature_group_id: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- ML Models & Registry Schemas ---
class MLModelBase(BaseModel):
    model_name: str = Field(..., min_length=1, max_length=255)
    algorithm: str = Field(..., min_length=1, max_length=100)
    framework: str = Field(default="scikit-learn", max_length=50)
    problem_type: str = Field(default="Classification", max_length=50)
    current_version: str = Field(default="v1.0.0", max_length=50)
    status: str = Field(default="Production", max_length=50)


class MLModelCreate(MLModelBase):
    organization_id: uuid.UUID


class MLModelResponse(MLModelBase):
    id: uuid.UUID
    organization_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ModelVersionCreate(BaseModel):
    model_id: uuid.UUID
    version: str = Field(..., min_length=1, max_length=50)
    metrics: str = Field(..., min_length=1)  # JSON string
    artifact_path: str = Field(..., min_length=1, max_length=500)


class ModelVersionResponse(BaseModel):
    id: uuid.UUID
    model_id: uuid.UUID
    version: str
    metrics: str
    artifact_path: str
    registered_at: datetime
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Training & Predictions Schemas ---
class TrainingJobCreate(BaseModel):
    model_id: uuid.UUID
    dataset_id: uuid.UUID


class TrainingJobResponse(BaseModel):
    id: uuid.UUID
    model_id: uuid.UUID
    dataset_id: uuid.UUID
    started_at: datetime
    completed_at: datetime | None = None
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class PredictionRequest(BaseModel):
    model_version_id: uuid.UUID
    input_data: dict = Field(...)


class PredictionResult(BaseModel):
    prediction_id: uuid.UUID
    model_version_id: uuid.UUID
    prediction_output: dict
    latency_sec: float
    confidence_score: float
    status: str


# --- Drift Monitoring Schemas ---
class DriftReportCreate(BaseModel):
    model_id: uuid.UUID
    drift_type: str = Field(default="Data Drift", max_length=50)


class DriftReportResponse(BaseModel):
    id: uuid.UUID
    model_id: uuid.UUID
    drift_type: str
    drift_score: float
    threshold: float
    status: str
    generated_at: datetime
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# --- Dashboard Summary Schema ---
class DataPlatformDashboardSummary(BaseModel):
    total_datasets: int
    total_pipeline_runs: int
    total_features_in_store: int
    active_ml_models: int
    total_predictions: int
    active_drift_alerts: int
    average_prediction_latency_ms: float
    training_jobs_completed: int

File: apps/web/src/services/dataAnalyticsMlops.ts
import { apiClient } from './apiClient';

export interface Dataset {
  id: string;
  organization_id: string;
  dataset_name: string;
  dataset_type: string;
  source: string;
  schema_version: string;
  status: string;
  created_at: string;
}

export interface PipelineJob {
  id: string;
  organization_id: string;
  pipeline_name: string;
  schedule_cron?: string;
  status: string;
  created_at: string;
}

export interface ETLJob {
  id: string;
  job_name: string;
  source_type: string;
  target_type: string;
  status: string;
  created_at: string;
}

export interface FeatureStore {
  id: string;
  feature_name: string;
  feature_group: string;
  data_type: string;
  description?: string;
  version: string;
  status: string;
  created_at: string;
}

export interface MLModel {
  id: string;
  organization_id: string;
  model_name: string;
  algorithm: string;
  framework: string;
  problem_type: string;
  current_version: string;
  status: string;
  created_at: string;
}

export interface ModelVersion {
  id: string;
  model_id: string;
  version: string;
  metrics: string;
  artifact_path: string;
  registered_at: string;
}

export interface PredictionResult {
  prediction_id: string;
  model_version_id: string;
  prediction_output: any;
  latency_sec: number;
  confidence_score: number;
  status: string;
}

export interface DriftReport {
  id: string;
  model_id: string;
  drift_type: string;
  drift_score: number;
  threshold: number;
  status: string;
  generated_at: string;
}

export interface DataPlatformDashboardSummary {
  total_datasets: number;
  total_pipeline_runs: number;
  total_features_in_store: number;
  active_ml_models: number;
  total_predictions: number;
  active_drift_alerts: number;
  average_prediction_latency_ms: number;
  training_jobs_completed: number;
}

export const dataAnalyticsMlopsService = {
  // Dashboard
  getDashboardSummary: async (orgId: string): Promise<DataPlatformDashboardSummary> => {
    const res = await apiClient.get('/api/v1/analytics/dashboard', { params: { org_id: orgId } });
    return res.data;
  },

  // Datasets
  getDatasets: async (orgId: string): Promise<Dataset[]> => {
    const res = await apiClient.get('/api/v1/analytics/datasets', { params: { org_id: orgId } });
    return res.data;
  },

  createDataset: async (data: Partial<Dataset>): Promise<Dataset> => {
    const res = await apiClient.post('/api/v1/analytics/datasets', data);
    return res.data;
  },

  // Pipelines & ETL
  getPipelines: async (orgId: string): Promise<PipelineJob[]> => {
    const res = await apiClient.get('/api/v1/analytics/pipelines', { params: { org_id: orgId } });
    return res.data;
  },

  createPipeline: async (data: Partial<PipelineJob>): Promise<PipelineJob> => {
    const res = await apiClient.post('/api/v1/analytics/pipelines', data);
    return res.data;
  },

  getETLJobs: async (): Promise<ETLJob[]> => {
    const res = await apiClient.get('/api/v1/analytics/etl-jobs');
    return res.data;
  },

  createETLJob: async (data: Partial<ETLJob>): Promise<ETLJob> => {
    const res = await apiClient.post('/api/v1/analytics/etl-jobs', data);
    return res.data;
  },

  // Feature Store
  getFeatures: async (): Promise<FeatureStore[]> => {
    const res = await apiClient.get('/api/v1/analytics/features');
    return res.data;
  },

  createFeature: async (data: Partial<FeatureStore>): Promise<FeatureStore> => {
    const res = await apiClient.post('/api/v1/analytics/features', data);
    return res.data;
  },

  // ML Models & Registry
  getModels: async (orgId: string): Promise<MLModel[]> => {
    const res = await apiClient.get('/api/v1/analytics/models', { params: { org_id: orgId } });
    return res.data;
  },

  registerModel: async (data: Partial<MLModel>): Promise<MLModel> => {
    const res = await apiClient.post('/api/v1/analytics/models', data);
    return res.data;
  },

  // Prediction Sandbox
  runOnlinePrediction: async (modelVersionId: string, inputData: any): Promise<PredictionResult> => {
    const res = await apiClient.post('/api/v1/analytics/predictions/online', {
      model_version_id: modelVersionId,
      input_data: inputData,
    });
    return res.data;
  },

  // Drift Monitoring
  generateDriftReport: async (modelId: string, driftType: string): Promise<DriftReport> => {
    const res = await apiClient.post('/api/v1/analytics/drift/reports', {
      model_id: modelId,
      drift_type: driftType,
    });
    return res.data;
  },
};
